src/gt4py/next/iterator/type_inference.py

```diff
@@ -380,20 +385,54 @@ def _infer_shift_location_types(shift_args, offset_provider, constraints):
     return current_loc_in, current_loc_out
 
 
-class _TypeInferrer(eve.NodeTranslator):
+@dataclasses.dataclass
+class _TypeInferrer(eve.traits.VisitorWithSymbolTableTrait, eve.NodeTranslator):
     """Visit the full iterator IR tree, convert nodes to respective types and generate constraints."""
 
-    def __init__(self, offset_provider):
-        self.offset_provider = offset_provider
+    offset_provider: Optional[dict[str, Connectivity | Dimension]]
+    collected_types: dict[int, Type] = dataclasses.field(default_factory=dict)
+    constraints: set[tuple[Type, Type]] = dataclasses.field(default_factory=set)
```

Comment:
For future reference, could you add a short description of how the information is encoded in these types?

Reply:
I did. Not super happy with the clarity of them though.